deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/FancyBlockingQueue.java

```diff
@@ -0,0 +1,288 @@
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.LockSupport;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * This BlockingQueue implementation is suited only for symmetric gradients updates, and should NOT be used anywhere else.
+ *
+ * Basic idea: all worker threads requesting via poll()/take() method will be advancing only once all consumers get the same element from Queue.
+ * So, multiple consumers are guaranteed to be consuming the same elements in the same order served by this queue.
+ *
+ * @author raver119@gmail.com
+ */
+@Slf4j
+public class FancyBlockingQueue<E> implements BlockingQueue<E>, Registerable {
+    protected BlockingQueue<E> backingQueue;
+    protected volatile int consumers;
+
+    protected ThreadLocal<AtomicLong> currentStep = new ThreadLocal<>();
+    protected final AtomicLong step = new AtomicLong(0);
+    protected final AtomicInteger state = new AtomicInteger(0);
+    protected final AtomicInteger currentConsumers = new AtomicInteger(0);
+
+    protected AtomicBoolean isFirst = new AtomicBoolean(false);
+    protected AtomicBoolean isDone = new AtomicBoolean(true);
+
+    protected AtomicInteger barrier = new AtomicInteger(0);
+    protected AtomicInteger secondary = new AtomicInteger(0);
+
+    protected AtomicInteger numElementsReady = new AtomicInteger(0);
+    protected AtomicInteger numElementsDrained = new AtomicInteger(0);
+    protected AtomicBoolean bypassMode = new AtomicBoolean(false);
+
+    protected boolean isDebug = false;
+    protected ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+
+
+    public FancyBlockingQueue(@NonNull BlockingQueue<E> queue){
+        this(queue, -1);
+    }
+
+    public FancyBlockingQueue(@NonNull BlockingQueue<E> queue, int consumers) {
+        this.backingQueue = queue;
+        this.consumers = consumers;
+        this.currentConsumers.set(consumers);
+    }
+
+
+    @Override
+    public boolean add(E e) {
+        return backingQueue.add(e);
+    }
+
+    @Override
+    public boolean offer(E e) {
+        return backingQueue.offer(e);
+    }
+
+
+    @Override
+    public E remove() {
+        return backingQueue.remove();
+    }
+
+    @Override
+    public void fallbackToSingleConsumerMode(boolean reallyFallback) {
+        bypassMode.set(reallyFallback);
+    }
+
+    @Override
+    public void registerConsumers(int consumers) {
+        lock.writeLock().lock();
+
+        this.numElementsReady.set(backingQueue.size());
+        this.numElementsDrained.set(0);
+        this.consumers = consumers;
+        this.currentConsumers.set(consumers);
+
+        lock.writeLock().unlock();
+    }
+
+    @Override
+    public void put(E e) throws InterruptedException {
+        lock.readLock().lock();
+        backingQueue.put(e);
+        lock.readLock().unlock();
```

Comment:
That's not for queue. That's to ensure that all local consumers will have exactly the same amount of external gradients applied, since queue advances for everyone at the same time. CUDA-style parallelism.

And read locks are cheap :)